### ISSUE TITLE
[Enhancement] Sampling from discrete distributions with integer index instead of 1-hot

### ISSUE THREAD

**Issue by dwf** [Theano/Theano#2759]
If you want to sample from a very high cardinality discrete distribution (or draw a lot of samples from smaller ones), you may run into memory problems using `multinomial` if you just want the indices of the sampled value. In any case, constructing the one-hot array only to take the argmax on axis 1 is very inefficient. It would be nice to have a "discrete" method for rng_mrg which just gave you the index of the sampled value (although, can GPU ops have integer outputs yet?).

**Comment by lamblin** [comment]
So, it would be like `random_integers`, but with a probability vector like the one passed to `multinomial`, right?
Integer outputs are not handled by the current back-end, but the new one will.

**Comment by lamblin** [comment]
Actually, `choice` seems to do that for the regular randomstreams.

**Comment by dwf** [comment]
If I'm not mistaken, choice() is limited to a single distribution being
specified  in pvals whereas multinomial can handle a matrix of pvals.

**Comment by lamblin** [comment]
Right. We could expand `choice` to support a tensor of values, like we did for `multinomial` in the first place.
